man/cq: Document range for signaling vector

shefty opened the pull request:
Identify the range for signaling vectors and indicate
that the value may be ignored by the provider.  Update
the utility code to accept any value, but still print
a warning if it is non-zero.

Fixes #778.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

bgoglin commented:
What's a "processor core"? You're likely talking about logical processor/hardware thread/hyperthread instead. Core numbers are different, and they are not unique in multi-socket x86 hosts.

shefty commented:
I'm referring to a logical processor thingy.  :)  We need to determine if this is referring to a CPU or a CPU + core or CPU + core + thread, or if these options are even possible.

shefty commented:
Carrying over the conversation from the previous PR, these are the alternatives I can think of:

1. CPU is 1-based index
2. CPU is 0-based index, with a new flag added to indicate when the index is valid
3. signaling_vector is an encoding of some bit fields

Option 3 can include being option 1 or 2 (for backwards compatibility).  It only acts as a bit field if the high-order bit is set.

jsquyres commented:
I think @bgoglin is right.  Isn't the correct terminology "Linux virtual processor ID"?

shefty commented:
FWIW, here's what hwloc defines for 'cpus':
http://linux.die.net/man/7/hwloc

node
A NUMA node; a set of processors around memory which the processors can directly access.

socket
Typically a physical package or chip, it is a grouping of one or more processors.

core
A single, physical processing unit which may still contain multiple logical processors, such as hardware threads.

pu
Short for processor unit (not process!). The smallest physical execution unit that hwloc recognizes. For example, there may be multiple PUs on a core (e.g., hardware threads).

The concept of breaking this down into bit fields _could_ allow us to get into this level of details... somewhat

shefty commented:
Updated text to "logical cpu number" (taken from lscpu man page).  But this doesn't mean that I think this is the right solution.

shefty commented:
I updated the PR.  It turns out that the verbs provider passes the signaling_vector value through to verbs, which expects a 0-based index.  As @j-xiong pointed out, a 0-based value is more in line with Linux.  So I ended up going with option 2 mentioned above.  I added a new flag -- FI_AFFINITY -- that indicates when signaling_vector is valid.

For the verbs provider, the signaling_vector value continues to be passed through to verbs, which leaves the behavior of any apps using this feature unchanged.  Since no other providers support signaling_vector, adding the flag also leaves existing behavior unchanged.  :)

Using a flag gives us more flexibility in the future, as different flags could be used to change how the provider interprets the signaling_vector.  E.g. we could add FI_AFFINITY_MASK if we want signaling_vector to be used as a bitmask, rather than an index.  We can consider converting signaling_vector from int to u64, or similar, as part of an ABI bump.

goodell commented:
I'm happy with this approach. 👍 

I'm also in favor of bumping the `signaling_vector` up to `u64` when we bump the ABI.  That would make it easier to do things like pass a `cpu_set_t *` at some point if we wanted.

j-xiong commented:
Looks good. 👍
